### PR TITLE
[WC_Tracker] Support for get_tracking_history

### PR DESCRIPTION
**Pull request by puntope**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

THis PR adds new data into the Tracker Snapshots:

- woocommerce_allow_tracking ( 'yes' or 'no' ) - If woocommerce_allow_tracking is enabled.
- woocommerce_allow_tracking_last_modified (timestamp / 'unknown' ) - Last time the option was changed.
- woocommerce_allow_tracking_first_optin (timestamp  / 'unknown' ) - First time the option was set as 'yes".

PT: pffQ75-1bu-p2

### How to test the changes in this Pull Request:

1. Checkout this PR
2. Run `pnpm -- wp-env run cli -- wp wc tracker snapshot --format=yaml`
3. See the new values at the end of the data stream
4. The expected values should be:
```
  woocommerce_allow_tracking: 'no' 
  woocommerce_allow_tracking_last_modified: "unknown"
  woocommerce_allow_tracking_first_optin: "unknown"
```
5. Run `pnpm -- wp-env run cli -- wp option update woocommerce_allow_tracking 'yes'``
6. Run `pnpm -- wp-env run cli -- wp wc tracker snapshot --format=yaml`
7. Lets assume 123455678 is the current timestamp. Then the expected values should be :
```
  woocommerce_allow_tracking: 'yes' 
  woocommerce_allow_tracking_last_modified: 123455678
  woocommerce_allow_tracking_first_optin:  123455678
```
8. Run `pnpm -- wp-env run cli -- wp option update woocommerce_allow_tracking 'no'`
9.  Run `pnpm -- wp-env run cli -- wp wc tracker snapshot --format=yaml`
10. Lets assume 123123123 is the current timestamp. Then the expected values should be :
```
  woocommerce_allow_tracking: 'no' 
  woocommerce_allow_tracking_last_modified: 123123123
  woocommerce_allow_tracking_first_optin:  123455678
```
11. Run `pnpm -- wp-env run cli -- wp option update woocommerce_allow_tracking 'yes'`
12.  Run `pnpm -- wp-env run cli -- wp wc tracker snapshot --format=yaml`
13. Lets assume 87654321 is the current timestamp. Then the expected values should be :
```
  woocommerce_allow_tracking: 'yes' 
  woocommerce_allow_tracking_last_modified: 87654321
  woocommerce_allow_tracking_first_optin:  123455678
```
14. Clear the options and test again using UI.  (Setting the Opt-in in WooCommerce Onboarding and Changing the values in WooCommerce -> Settings -> Advanced -> WooCommerce.com -> "Allow usage of WooCommerce to be tracked")

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/includes/class-woocommerce.php [hunk 1]
```diff
@@ -1341,4 +1342,25 @@ public function register_remote_log_handler( $handlers ) {
...
+	 * - When the field was first set to allow tracking
+	 * - Last time the option was changed
+	 *
+	 * @param string $old_value The old value for the woocommerce_allow_tracking option.
+	 * @param string $value The current value for the woocommerce_allow_tracking option.
+	 * @since x.x.x
+	 *
+	 * @return void
+	 */
+	public function get_tracking_history( $old_value, $value ) {
```
Comment (mikkamp) [inline plugins/woocommerce/includes/class-woocommerce.php:1357]: Looking at the contents of this function I only see it calling two "update" functions. It's also hooked into an "update" hook. Shouldn't the function name reflect the same thing, that it's updating the tracking history?

File: plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php [hunk 2]
```diff
@@ -305,4 +305,54 @@ public function test_get_tracking_data_admin_install_timestamp() {
...
+
+		sleep( 1 ); // be sure $time_three is at least one second after $time_two.
+		$time_three = time();
+		update_option( 'woocommerce_allow_tracking', 'yes' );
+		$tracking_data = WC_Tracker::get_tracking_data();
+		$this->assertEquals( $tracking_data['woocommerce_allow_tracking'], 'yes' );
+		$this->assertTrue( $tracking_data['woocommerce_allow_tracking_last_modified'] >= $time_three );
+		$this->assertTrue( $tracking_data['woocommerce_allow_tracking_first_optin'] >= $time_one && $tracking_data['woocommerce_allow_tracking_first_optin'] < $time_two );
+
+		// Restore everything as it was.
```
Comment (mikkamp) [inline plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php:353]: Just a small nitpick. We are restoring `woocommerce_allow_tracking`, but if the other two were set before this test then it's not really restoring them. Doesn't really change the tests in any way though.

File: plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php [hunk 3]
```diff
@@ -305,4 +305,54 @@ public function test_get_tracking_data_admin_install_timestamp() {
...
+
+		$this->assertEquals( $tracking_data['woocommerce_allow_tracking'], 'no' );
+		$this->assertEquals( $tracking_data['woocommerce_allow_tracking_last_modified'], 'unknown' );
+		$this->assertEquals( $tracking_data['woocommerce_allow_tracking_first_optin'], 'unknown' );
+
+		$time_one = time();
+		update_option( 'woocommerce_allow_tracking', 'yes' );
+		$tracking_data = WC_Tracker::get_tracking_data();
+		$this->assertEquals( $tracking_data['woocommerce_allow_tracking'], 'yes' );
+		$this->assertTrue( $tracking_data['woocommerce_allow_tracking_last_modified'] >= $time_one );
```
Comment (mikkamp) [inline plugins/woocommerce/tests/php/includes/class-wc-tracker-test.php:333]: `>=` is a very broad range to check against. I understand we don't want to introduce any flakiness, if it jumps to the next second right in the middle of this test. But should we add an additional check for `< $time_one + MINUTE_IN_SECONDS`?

**Review by mikkamp — APPROVED** [review]
Thanks for the changes here. The extra snapshot data is being populated as expected:

```
woocommerce_allow_tracking: 'yes'
woocommerce_allow_tracking_last_modified: "1735572611"
woocommerce_allow_tracking_first_optin: "1735572548"
```

I just added a few small comments as I looked through the code changes, but none of those block the functionality of this PR so I'll go ahead and pre-approve it.

**Comment by puntope** [comment]
Thanks @mikkamp For the review. I will add the points mentioned in a follow up PR.

**Pull request merged**